src/main/java/org/folio/inventory/consortium/services/ConsortiumServiceImpl.java

```diff
@@ -0,0 +1,131 @@
+
+public class ConsortiumServiceImpl implements ConsortiumService{
+  private static final Logger LOGGER = LogManager.getLogger(ConsortiumServiceImpl.class);
+  private static final String USER_TENANTS_ENDPOINT = "/user-tenants?limit=1";
+  private static final String SHARE_INSTANCE_ENDPOINT = "/consortia/%s/sharing/instances";
+  private static final String CONSORTIA_ENDPOINT = "/consortia";
+  private static final String SHARING_INSTANCE_ERROR = "Error during sharing Instance for sourceTenantId: %s, targetTenantId: %s, instanceIdentifier: %s, status code: %s, response message: %s";
+  private final HttpClient httpClient;
+
+  public ConsortiumServiceImpl(HttpClient httpClient) {
+    this.httpClient = httpClient;
+  }
+
+  public Future<SharingInstance> createShadowInstance(Context context, String instanceId) {
+    return getCentralTenantId(context)
+      .compose(centralTenantId -> {
+        Context centralTenantContext = constructContext(centralTenantId, context.getToken(), context.getOkapiLocation());
+        return getConsortiumId(centralTenantContext)
+          .compose(consortiumId -> {
+            SharingInstance sharingInstance = new SharingInstance();
+            sharingInstance.setSourceTenantId(centralTenantId);
+            sharingInstance.setInstanceIdentifier(UUID.fromString(instanceId));
+            sharingInstance.setTargetTenantId(context.getTenantId());
+            return shareInstance(centralTenantContext, consortiumId, sharingInstance);
+          });
+      });
+  }
+
+  public Future<String> getCentralTenantId(Context context) {
+    CompletableFuture<String> completableFuture = createOkapiHttpClient(context)
+      .thenCompose(client ->
+        client.get(context.getOkapiLocation() + USER_TENANTS_ENDPOINT).toCompletableFuture()
+          .thenCompose(httpResponse -> {
+            if (httpResponse.getStatusCode() == HttpStatus.SC_OK) {
+              JsonArray userTenants = httpResponse.getJson().getJsonArray("userTenants");
+              if (userTenants.isEmpty()) {
+                String message = "Central tenant not found";
+                LOGGER.warn(String.format("getCentralTenantId:: %s", message));
+                return CompletableFuture.failedFuture(new NotFoundException(message));
+              }
+              String centralTenantId = userTenants.getJsonObject(0).getString("centralTenantId");
+              LOGGER.debug("getCentralTenantId:: Found centralTenantId: {}", centralTenantId);
+              return CompletableFuture.completedFuture(centralTenantId);
+            } else {
+              String message = String.format("Error retrieving centralTenantId by tenant id: %s, status code: %s, response message: %s",
+                context.getTenantId(), httpResponse.getStatusCode(), httpResponse.getBody());
+              LOGGER.warn(String.format("getCentralTenantId:: %s", message));
+              return CompletableFuture.failedFuture(new EventProcessingException(message));
```

Comment:
Not sure here, but EventProcessingException looks like some exception during event processing. So if this service is going to be reused somewhere else, like within rest endpoint, it might look misleading

Reply:
Thanks, done